Added loop to iterate to vehicle tile_element before checking cover.

martip23 opened the pull request:
Added loop to iterate to vehicle tile_element before checking if covered above. Fixes #7405

IntelOrca commented on src/openrct2/ride/Vehicle.cpp:
```diff
@@ -1830,9 +1831,26 @@ static void vehicle_update_measurements(rct_vehicle * vehicle)
+    if (tile_element->base_height * 8 <= z) // If vehicle above ground.
     {
+        // Couldn't get this to work?
+        //tile_element = map_get_nth_element_at(x/32, y/32, z/8);
```
The third parameter should be an index not a height.

Broxzier commented on src/openrct2/ride/Vehicle.cpp:
```diff
@@ -1830,9 +1831,26 @@ static void vehicle_update_measurements(rct_vehicle * vehicle)
 
     rct_tile_element * tile_element = map_get_surface_element_at({x, y});
-    if (tile_element->base_height * 8 <= vehicle->z)
+    if (tile_element->base_height * 8 <= z) // If vehicle above ground.
```
We place comments above statements.

Broxzier commented on src/openrct2/ride/Vehicle.cpp:
```diff
@@ -1830,9 +1831,26 @@ static void vehicle_update_measurements(rct_vehicle * vehicle)
+                    tile_element_at_ride_height = true;
+                }
+            }
+        } while (!tile_element_at_ride_height);
```
The logic here has a few flaws in it. When there are multiple track elements of the same ride on the same tile, it will always find the first element. The second flaw is that tile elements aren't necessarily ordered from bottom to top (though this is not achievable through normal gameplay), meaning the next loop will start iterating at a seemingly random element.

Instead of picking the tile element to start iterating from, you should start from the first element on the tile and at the beginning of the loop compare its base height with the vehicle's height (keep in mind that `base_height` uses small coordinates, so multiply this with 8 first). If it's lower, `continue`. Otherwise the rest of the loop will executed and `cover_found` will be set correctly.

martip23 replied:
Thank you, I believe I've integrated the changes. Your comment was a great help.

Broxzier commented:
This PR looks much better now! Did you close this by accident?

martip23 commented:
I've closed it because I introduced a bug which broke the build by accidentally changing 1846. I've fixed it now. This is my first time working with a large team so I wasn't sure what the etiquette was. I closed it to prevent it from being merged just yet. It's ready to be merged now I believe, so I'm reopening.

![image](https://user-images.githubusercontent.com/16685010/39677504-87ae9e0e-5141-11e8-851a-c7a366704311.png)
Funny story... So I thought it was still borked during testing since people weren't getting on the ride despite the coverage. It turns out that people just didn't want to pay $1 for my badass train ride. Once I set it for 10 cents, people lined up and we can confirm that coverage is working properly for this situation.

Broxzier commented:
That's fair. There's not really an etiquette for unfinished PRs. Next time you can just mention it in a comment and we can add the work-in-progress label if necessary. Before PRs get merged we always try to test them thoroughly, and a cimpilation failure would have been caught by the CIs already.

Could you add a [changelog entry](https://github.com/OpenRCT2/OpenRCT2/blob/develop/distribution/changelog.txt) for #7405? They're ordered by issue number.

And if you're familiar enough with git, could you rebase and squash this to get rid of the merge commits?

AaronVanGeffen commented on distribution/changelog.txt:
```diff
@@ -1,4 +1,4 @@
-0.1.3 (in development)
+﻿0.1.3 (in development)
```
It looks like your editor added a byte order mark (BOM) here. Could you remove it?

Broxzier replied:
FWIW Notepad does this by default, Notepad++ and Visual Studio do not.

martip23 replied:
Thank you. Notepad decided to add it. I've removed it with VIM.

Broxzier commented on distribution/changelog.txt:
```diff
@@ -18,8 +18,7 @@
 - Fix: [#7402] Edges of neigbouring footpaths stay connected after removing a path that's underneath a ride entrance.
-- Fix: [#7405] Vehicle test now checks for cover only on tiles above
-vehicle.
+- Fix: [#7405] Vehicle test now checks for cover only on tiles above vehicle.
```
This should state the bug that has been fixed, rather than the change that has been made.

Broxzier replied:
I probably wasn't clear enough, what I mean is an entry like `- Fix: [#7405] Rides can be covered by placing scenery underneath them.`

martip23 replied:
Done.

martip23 commented:
I am not quite sure how to use rebase. It seems I've just ended up making more unnecessary commits. I will just push it as is to get the changes passed and work more on my git-fu some other time.

Broxzier commented:
I've rebased for you and took it for a test, and everything seems to work as expected. Now when CI is green this can be merged.

martip23 commented:
My first contribution to an open source project 😁
